src/zc/zodbdgc/README.test

```diff
@@ -726,6 +726,219 @@ files directly:
+
+Next, we add a persistent object to the first database:
+
+    >>> conn1 = db.open()
+    >>> conn2 = conn1.get_connection('db2')
+    >>> pers = conn1.root.x = C()
+    >>> conn1.root.x_alias = pers
+    >>> conn1.add(pers)
+
+We can make a weak reference to this object in both databases
+as well as a strong cross-DB reference:
+
+    >>> from persistent.wref import WeakRef
+    >>> conn1.root.wx = WeakRef(pers)
+    >>> conn2.root.wx = WeakRef(pers)
+    >>> transaction.commit()
+    >>> conn2.root.x = pers
+    >>> transaction.commit()
+
+Time passes. :)
+
+    >>> now += 7 * 86400        # 7 days
+
+The number of objects in the databases now:
+
+    >>> len(conn1._storage), len(conn2._storage)
+    (2, 1)
+
+Packing doesn't change it:
+
+    >>> for d in db.databases.values():
+    ...     d.pack()
+    >>> len(conn1._storage), len(conn2._storage)
+    (2, 1)
+    >>> _ = [d.close() for d in db.databases.values()]
+
+We can GC and nothing is collected:
+
+    >>> zc.zodbdgc.gc('config', days=2)
+    db1: roots
+    db1: recent
+    db2: roots
+    db2: recent
+    db1: remove garbage
+    Removed 0 objects from db1
+    db2: remove garbage
+    Removed 0 objects from db2
+
+Now we can delete the object and jump ahead:
+
+    >>> with open('config', 'r') as f:
+    ...     db = ZODB.config.databaseFromFile(f)
+    >>> conn1 = db.open()
+    >>> conn2 = conn1.get_connection('db2')
+    >>> del conn1.root.x
+    >>> transaction.commit()
+    >>> now += 7 * 86400        # 7 days
+    >>> len(conn1._storage), len(conn2._storage)
+    (2, 1)
+    >>> db.pack()
+    >>> len(conn1._storage), len(conn2._storage)
+    (2, 1)
+    >>> _ = [d.close() for d in db.databases.values()]
+
+Still no garbage (the alias holds onto it) is collected:
+
+    >>> now += 7 * 86400        # 7 days
+    >>> zc.zodbdgc.gc('config', days=2)
+    db1: roots
+    db1: recent
+    db2: roots
+    db2: recent
+    db1: remove garbage
+    Removed 0 objects from db1
+    db2: remove garbage
+    Removed 0 objects from db2
+
+The weakrefs still work until we remove the alias:
+
+    >>> with open('config', 'r') as f:
+    ...     db = ZODB.config.databaseFromFile(f)
+    >>> conn1 = db.open()
+    >>> conn2 = conn1.get_connection('db2')
+    >>> conn1.root.wx() is not None
+    True
+    >>> conn2.root.wx() is not None
+    True
+    >>> del conn1.root.x_alias
+    >>> transaction.commit()
+    >>> db.pack() #for d in db.databases.values(): d.pack()
+    >>> len(conn1._storage), len(conn2._storage)
+    (2, 1)
+    >>> _ = [d.close() for d in db.databases.values()]
+
+Now a GC will collect things from the database that the original
+object was in:
+
+    >>> now += 7 * 86400        # 7 days
+    >>> zc.zodbdgc.DEBUG = True
+    >>> zc.zodbdgc.gc('config', days=2, return_bad=True)
+    db1: roots
+    db1: recent
+    db2: roots
+    db2: recent
+    db1: remove garbage
+    Removed 1 objects from db1
+    db2: remove garbage
+    Removed 0 objects from db2
+    [('db1', 1)]
```

Comment:
Sadly this looks like a bug in the gc algorithm. The item in question, OID 1 in db1, is found in the Bad object after all references have been collected, so it is removed. I added some logging to both `oidset.insert` and `Bad.insert` to see why this would be. Here's the logging for this example:
```
    db1: roots
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01',)] db1 ()
    db1: recent
    gc_: Has data <ZODB.FileStorage.FileStorage.Record object at 0x10384bc10> 'cpersistent.mapping\nPersistentMapping\nq\x01.}q\x02U\x04dataq\x03}q\x04s.'
    BAD: Inserting db1 '\x00\x00\x00\x00\x00\x00\x00\x01' '\x03}\xfe\xe7\\8F\x99' <generator object getrefs at 0x103852780>
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01',)] db1 ()
    db2: roots
    ['m', ('db1', '\x00\x00\x00\x00\x00\x00\x00\x01', None)] db2 ()
    GOOD Insert db1 '\x00\x00\x00\x00\x00\x00\x00\x01'
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01', 'db1')] db2 ()
    db2: recent
    ['m', ('db1', '\x00\x00\x00\x00\x00\x00\x00\x01', None)] db2 ()
    GOOD Insert db1 '\x00\x00\x00\x00\x00\x00\x00\x01'
    ['w', ('\x00\x00\x00\x00\x00\x00\x00\x01', 'db1')] db2 ()
    db1: remove garbage
    Remove '\x00\x00\x00\x00\x00\x00\x00\x01' '\x03}\xfe\xe7\\8F\x99' db1
    Removed 1 objects from db1
    db2: remove garbage
    Removed 0 objects from db2
    [('db1', 1)]
```

You can see that OID 1 is inserted into Bad, although it is later inserted into the good oidset. Because it's found in Bad, it is removed.

Here's the relevant code from the `gc_` function that prints "Has data" and inserts the item into Bad:

```python
        for trans in it:
            for record in trans:
                ...

                oid = record.oid
                data = record.data
                if data:
                    if deleted.has(name, oid):
                        continue
                    if good.has(name, oid):
                       ... elide 2...
                    else:
                        if DEBUG and oid == '\x00\x00\x00\x00\x00\x00\x00\x01':
                            print("Has data", record, repr(data))
                        bad.insert(name, oid, record.tid,
                                   getrefs(data, name, ignore))

                else:
                    # deleted record
                    ...
```

In other words, we encounter the record in DB1 for OID 1. It's not deleted, and it's not known to be good (have other references), so it must be Bad. Even though traversing through DB2 later reveals it to be good, it never gets removed from Bad. 

I *think* this is because there are no references in its "home" database DB1, so it never gets put into the good set in the first place, and so the code that wants to pop from bad (... elide 2 ..., about line 278) never gets a chance to do so. But I'm not entirely certain.